Peer Review for my BucketList Application API

BolajiOlajide opened the pull request:
## What does this PR do?

It serves as my first output for Checkpoint 2

## Description of Task to be completed?

I am to create a Flask API for a bucket list service. Specification for the API is shown below. You may use any database you prefer for this assignment.

| EndPoint                                 | Functionality                 | Public Access|
| -----------------------------------------|:-----------------------------:|-------------:|
| **POST** /auth/register                  | Register a user               |    TRUE      |
| **POST** /auth/login                     | Logs a user in                |    TRUE      |
| **POST** /bucketlists/                   | Create a new bucket list      |    FALSE     |
| **GET** /bucketlists/                    | List all created bucket lists |    FALSE     |
| **GET** /bucketlists/id                  | Get single bucket list        |    FALSE     |
| **PUT** /bucketlists/id                  | Update a bucket list          |    FALSE     |
| **DELETE** /bucketlists/id               | Delete a bucket list          |    FALSE     |
| **POST** /bucketlists/id                   | Create a new item bucket list |    FALSE     |
| **PUT** /bucketlists/id/items/item_id      | Update a bucket list item     |    FALSE     |
| **DELETE** /bucketlists/id/items/item_id   | Delete an item in bucket list |    FALSE     |

## How should this be manually tested?

Instructions have been included in the README file.

## Screenshot

This is a view of how to use the API using POSTMAN.

<img width="988" alt="screen shot 2017-03-31 at 6 46 11 pm" src="https://cloud.githubusercontent.com/assets/25608335/24562591/28c6e8ae-1643-11e7-90cc-7b04bbc76822.png">

BolajiOlajide committed:
[chore] Add tests

Test the application for sad and happy paths

[CHORE] Edit test docstrings

Make changes to the docstrings of the tests and also add little changes to the test names.

[CHORE] Add tests

Test the application for sad and happy paths#

BolajiOlajide committed:
[chore] Configure Travis

Configure travis CI for build testing

BolajiOlajide committed:
[chore] Add requirements files

Add the requirements.txt files for developers and end users.

BolajiOlajide committed:
[chore] add coveralls files

add the .coveralls.yml and .coveragerc for communication with coveralls.io and coverage calculation

BolajiOlajide committed:
[chore] Add configuration script

add the config.py script to define the different configurations to be used within the application

BolajiOlajide committed:
[chore] Add skeleton files

Add README and license files.

[chore] Add migration details

Add the alembic migration initialization file to help with easy migration for the user.

[chore] Add skeleton files

Add README and license files and also Add the alembic migration initialization file to help with easy migration for the user.#

BolajiOlajide committed:
[chore] Add skeleton files

Add README and license files.

[chore] Configure Travis

Configure travis CI for build testing

[chore] Add requirements files

Add the requirements.txt files for developers and end users.

[chore] add coveralls files

add the .coveralls.yml and .coveragerc for communication with coveralls.io and coverage calculation

[CHORE] Add Flask WTF to requirements list

[chore] Squash commits#

BolajiOlajide committed:
[FEATURE] Setup App config and management

Make changes to the config file to be able to run in development, testing and production environments. Also set up the manage.py to be able to scale the management of the BucketList app.

BolajiOlajide committed:
[feature] Make a shell context for the app

Create a context accessible via an interactive python shell for the user and all ignore the environment variable file and .DS_Store.

BolajiOlajide committed:
[FEATURE] Add routes and blueprints for the API

Setup the routes, blueprints and models to be used by the API.

BolajiOlajide committed:
[CHORE] Correct config for TRavis Build

BolajiOlajide committed:
[CHORE] Edit nosetest config

BolajiOlajide committed:
[FEATURE] Implement the Pagination feature

Set config variables for pagination and also a decorator to be used for pagination

BolajiOlajide committed:
[chore] Use single requirements file

Delete the requirements folder that was housing two requirements.txt files and create a single one

BolajiOlajide committed:
[chore] add SERVER_NAME configuration

Specify the SERVER_NAME for the API, set it to localhost:5000

BolajiOlajide committed:
[chore] Delete helper file

BolajiOlajide committed:
[rf] Refactor various route functions

Refactor the different functions in the main route and also decorate the functions with the auth.login_required decorator so as to all for only authenticated users

BolajiOlajide committed:
[chore] Remove debug statements

BolajiOlajide committed:
[refactor] Use the current time for date_created

Use the current time for the variables date_created and date_modified, also increase token expiration to 36000seconds.

BolajiOlajide committed:
[feature] Add 401 error handler

BolajiOlajide committed:
[bug] Handle authentication error

Auth errors previously didn't return a json response, handle that with the new auth error handler

BolajiOlajide committed:
[ft] add status code for user login

BolajiOlajide committed:
[chore] Remove redundant route

Remove the token route for generating tokens, as the login route already handles that

BolajiOlajide committed:
Fix unauthorized handler bug

BolajiOlajide committed:
Ignore .env and test database

BolajiOlajide committed:
Add gunicorn to requirements

BolajiOlajide committed:
Add Procfile

BolajiOlajide committed:
chore-add database init command

erika-dike commented on app/auth/__init__.py:
```diff
@@ -0,0 +1,10 @@
+"""
+from flask import Blueprint
+
+authe = Blueprint('authe', __name__)
```
A more descriptive name would be helpful here.

BolajiOlajide replied:
I have changed the name to `authentication` instead to better describe what is being done.

erika-dike commented on app/auth/routes.py:
```diff
@@ -0,0 +1,95 @@
+@auth.verify_password
+def verify_password(token, password):
+    """
+    Verify token.
```
Why not call it verify_token instead? I think there is a class HTTPTokenAuth() that would serve you better.

BolajiOlajide replied:
I've checked out the HTTPTokenAuth() package. Seems it's a ruby package and not available for python.

I've changed the name of the method to verify_token instead.

erika-dike commented on app/auth/routes.py:
```diff
@@ -0,0 +1,95 @@
+    """
+    Verify a user's identity.
+
+    Verify the user's identity and returns a token.
```
Verify the user's identity is duplicated. The standard way to write a docstring can be found here:

1, https://www.python.org/dev/peps/pep-0008/#documentation-strings
2. https://www.python.org/dev/peps/pep-0257/

BolajiOlajide replied:
I have added aa better description to the docstring body.

So i was going through the documentation and turns out both styles of writing multi-line docstrings are equivalent. I usually start my doc string title on a new line because sometimes it exceeds 79 characters. So having the triple-quotes on a separate line makes room for 3 additional characters.

<img width="855" alt="screen shot 2017-04-06 at 7 26 04 am" src="https://cloud.githubusercontent.com/assets/25608335/24740700/740b7b3a-1a9a-11e7-86de-b15ff44a280d.png">

erika-dike commented on app/auth/routes.py:
```diff
@@ -0,0 +1,95 @@
+        return errors.bad_request("username or password missing.")
+
+    user = User.query.filter_by(username=username).first()
+    if not user or not user.verify_password(password):
```
Do you think an `and` is better here?

BolajiOlajide replied:
The problem with using an `and` is that the if statement only executes if both `not user` and `not user.verify_password(password)` conditions are true. In the even where `not user` is true and the second isn't, it'll not execute the statement. @andela-cdike 

Check this: [Repl.it snippet](https://repl.it/Gx9j/0).
When using an `and` statement, it evaluates to true if it encounters a true condition first.

BolajiOlajide replied:
I think  i've identified the problem, i have to use something like this:
`if not (user and user.verify_password(password)):`

erika-dike commented on app/auth/routes.py:
```diff
@@ -0,0 +1,95 @@
+    """
+    Create a new user.
+
+    Creates a new user in the application.
```
Same duplication

BolajiOlajide replied:
I've corrected this.

erika-dike commented on app/main/routes.py:
```diff
@@ -0,0 +1,217 @@
+        return errors.not_found("The BucketList with the id: {0} doesn't"
+                                " exist.".format(list_id))
+
+    if not request.json:
```
Don't you think this check should occur first? Since there will be no need querying the db if json isn't provided.

BolajiOlajide replied:
I've corrected this.

erika-dike commented on app/main/routes.py:
```diff
@@ -0,0 +1,217 @@
+        return errors.bad_request("No JSON file detected.")
+
+    if 'done' not in request.json or 'name' not in request.json:
+        return errors.bad_request("The keys 'done' and 'name' not in the JSON")
```
The message here is a bit misleading since it could either be `done` or `name` that isn't provided.

BolajiOlajide replied:
I've corrected this.

erika-dike commented on app/main/routes.py:
```diff
@@ -0,0 +1,217 @@
+        return errors.not_found("The item with the ID: {0} doesn't exist"
+                                .format(item_id))
+
+    if not request.json or (('name' or 'done') not in request.json):
```
This line of code is interesting. To me if feels there is bug somewhere: ('name' or 'done') not in request.json but I am not sure.

BolajiOlajide replied:
I've corrected this.

erika-dike commented on app/models.py:
```diff
@@ -0,0 +1,273 @@
+    """
+
+    __tablename__ = 'users'
+    user_id = db.Column(db.Integer, primary_key=True)
```
Wouldn't the name `id` be sufficient? You could have that in the Base model as well.

BolajiOlajide replied:
I didn't use the name `id` because it is a python keyword and it might cause some errors when i try to use it someplace else.

erika-dike commented on app/models.py:
```diff
@@ -0,0 +1,273 @@
+    """
+
+    __tablename__ = 'bucketlist'
+    bucketlist_id = db.Column(db.Integer, primary_key=True)
```
Same comment as that of `user_id`

BolajiOlajide replied:
Addressed above.

erika-dike commented on app/views/views.py:
```diff
@@ -0,0 +1,19 @@
+
+    This route is for the home page.
+    """
+    # return "<h1> App WORKS! </h1>"
```
Time to delete this :)

BolajiOlajide replied:
Deleted. It shouldn't even be there in the first place.
Clumsy me!

erika-dike commented on config.py:
```diff
@@ -0,0 +1,87 @@
+        This static method initializes the application using whatever
+        configuration the user has chosen.
+        """
+        pass
```
Do you still use this method?

BolajiOlajide replied:
Yes, i use it in the `app/__init__.py`. To initialize my config.

mfalade replied:
This `init_app` method is not needed

```python
app.config.from_object(<env>)
``` 
☝️ should suffice to initialize your config just like you have done [here](https://github.com/andela-bolajide/BucketListAPI/blob/develop/app/__init__.py#L21)

This method and the call to it is redundant and should be removed

erika-dike commented on config.py:
```diff
@@ -0,0 +1,87 @@
+    Attributes such as SECRET_KEY are the same no matter the platform used.
+    """
+
+    SECRET_KEY = 'u35dfh___vv@$%%jkdsvjkhb___jdfnv93043klm__vdmkll55vd----__'
```
I thought you were thinking about using environment variables. That would have been cool.

erika-dike replied:
Cool because I want to learn it.

BolajiOlajide replied:
Done. I made use of `python-dotenv`.

andela-mkamau commented on tests/test_models.py:
```diff
@@ -0,0 +1,104 @@
+            name="Jump off. a cliff"
+        )
+        self.assertIsInstance(item, Base)
+
```
I think the tests  `test_user_base_inheritance` , `test_bucketlist_base_inheritaance` and `test_items_base_inheritance` are unnecessary. Checking if one class is a subclass of another is testing Python libraries, not the functionality you introduced.

BolajiOlajide replied:
I'll work on that.

andela-mkamau commented on config.py:
```diff
@@ -0,0 +1,82 @@
+    'testing': TestingConfig,
+    'production': ProductionConfig,
+    'default': DevelopmentConfig
+}
```
I like the way you have structured your configurations. That's pretty sweet. To make it even better, you could put some config variables like secrets and database URLs in environment variables.

BolajiOlajide replied:
I've done that. Thanks Mike @andela-mkamau

andela-mkamau commented on app/auth/routes.py:
```diff
@@ -0,0 +1,102 @@
+
+    if not user.verify_password(password):
+        return errors.bad_request("Username and password doesn't match.")
+
```
For wrong username/password combination, I think you should return a 401-Unauthorised error not a 400 -Bad request. What do you think @andela-bolajide ?

BolajiOlajide replied:
You're absolutely right Mike.

andela-mkamau commented on app/main/routes.py:
```diff
@@ -0,0 +1,209 @@
+    bucketlist = BucketList.query.get_or_404(list_id)
+
+    if not bucketlist:
+        abort(404)
```
I think you shouldn't have this if statement here. `bucketlist` will never be a `None`, hence that `abort` will never be reached.  Check http://flask-sqlalchemy.pocoo.org/2.1/api/#flask.ext.sqlalchemy.BaseQuery.get_or_404 . Instead of `get_or_404`, just use `get`

BolajiOlajide replied:
Fixed.

andela-mkamau commented on app/main/routes.py:
```diff
@@ -142,21 +142,27 @@ def add_bucketlist_item(list_id):
 
-    if 'name' in request.json and type(request.json['name']) != unicode:
-        abort(400)
+    if not request.json or (('name' or 'done') not in request.json):
```
Love this line 🏆

BolajiOlajide replied:
Thank you. Mike

mfalade replied:
If you are trying to assert that either of those values exists in `request.json`, then this is a wrong approach.
```python
(('name' or 'done') not in request.json)

# evaluates to 

('name' not in request.json) # 'done' would never be checked and I don't think that's what you want

"""Because with the OR operator, the right value is evaluated ONLY when 
the left value is falsey.

So
"""

('' or 'foo') would evaluate to 'foo'
('foo' or '') would evaluate to 'foo'
(0 or 500) would evaluate to 500
(100 or 500) would evaluate to 100
(False or 'Not Good') would evaluate to 'Not Good'
(True or 'Python Rocks') would evaluate to True
('good' or 'bad') would evaluate to 'good'

```

andela-mkamau commented on requirements.txt:
```diff
@@ -3,6 +3,7 @@ coverage
+Flask-JWT
 Flask-Login
 Flask-Migrate
 Flask-Script
```
It's normally a good idea to have specific version numbers for your packages. Might prevent things from blowing 🔥 up when packages fail to play along.

BolajiOlajide replied:
Done

jugba reviewed (commented):
Where are the tests for your endpoints?

BolajiOlajide committed:
[tests] test the api functionality

add tests to test for sad paths and happy paths.

BolajiOlajide commented:
@andela-jugba Hello, i've added the tests. I haven't completed them that's why i didn't push them earlier.

BolajiOlajide committed:
chore: correct requirements.txt path

BolajiOlajide committed:
tests: test API endpoints

BolajiOlajide committed:
rf: refactor the API routes

Refactor the API routes and also correct some redundant errors

BolajiOlajide committed:
chore: delete template files

delete all files associated with the frontend view and static files too

BolajiOlajide committed:
chore: delete the views

BolajiOlajide committed:
chore: ignore test database

BolajiOlajide committed:
chore: add coveralls command

BolajiOlajide committed:
feature: implement environment variables

BolajiOlajide committed:
chore: delete redundant comment

BolajiOlajide committed:
rf: refactor tests

Redactor tests to read environment variables instead of the hard-coded variables.
Also add a test for the production config.

BolajiOlajide committed:
chore: add version number to requirements

BolajiOlajide committed:
chore: add python-dotenv package

BolajiOlajide committed:
create db tables using defined config

BolajiOlajide committed:
add coveralls to requirements

BolajiOlajide committed:
Remove codeclimate config and badge

BolajiOlajide committed:
Add .env sample template

BolajiOlajide committed:
edit heroku requirements

BolajiOlajide committed:
Update README with detailed information

BolajiOlajide committed:
Implement feedback gotten from PR

Changed the name of the authentication blueprint to a more descriptive one.
Change the name of the method that verifies token to better reflect it’s function.
Make docstrings more descriptive.

NdagiStanley commented on manage.py:
```diff
@@ -1,11 +1,69 @@
+Python script for managing the flask application.
 
-app = Flask(__name__)
+THis is the script that starts the flask application.
```
Nitpick *This

njirap replied:
hehe

NdagiStanley commented:
This being a solo project, I would advise not having `/migrations` on Github.
http://stackoverflow.com/questions/33784362/django-migrations-best-practice

BolajiOlajide committed:
Make use of werkzeug for password hash

BolajiOlajide committed:
Instantiate date variable on model class access

BolajiOlajide committed:
feature: handle database integrity error

BolajiOlajide committed:
feature: implement unique constraints

Make the name of a bucketlist and the person who created it unique, and also an item unique to a bucketlist

BolajiOlajide committed:
tests: correct test for identical name

BolajiOlajide committed:
Update README

BolajiOlajide committed:
Use environment variable for test

BolajiOlajide committed:
chore: remove debug statements

BolajiOlajide committed:
add flask to requirements

BolajiOlajide committed:
Configure Tox for Testing (#2)

* add tox config files

* add flask to requirements

* Ignore tox-related files

* resolve merge conflig

* Update README

Update the README with instructions on how to run the test using tox

* remove redundant files

* add tox to requirements.txt

* update readme

njirap commented on app/main/routes.py:
```diff
@@ -0,0 +1,218 @@
+        return errors.bad_request("Only JSON object is accepted. Please "
+                                  "confirm that the key 'name' exists.")
+
+    user = User.query.filter_by(username=g.user.username).first().user_id
```
isn't this just `g.user.user_id`?

BolajiOlajide replied:
I've corrected that.

njirap commented on app/main/routes.py:
```diff
@@ -0,0 +1,218 @@
+        bucketlist.save()
+    except (sqlalchemy.exc.IntegrityError, sqlalchemy.exc.InvalidRequestError):
+        db.session().rollback()
+        return errors.bad_request("An error occurred while saving. "
```
be more specific with error reports..

BolajiOlajide replied:
I will be more specific with my errors. I've fixed this message

njirap commented on app/main/routes.py:
```diff
@@ -0,0 +1,218 @@
+    Update a BucketList name.
+    """
+    if not request.json:
+        return errors.bad_request("No JSON file detected.")
```
are you really looking for a JSON file here?

BolajiOlajide replied:
Fixed.

BolajiOlajide replied:
I've rephrased my error message.

njirap commented on app/main/routes.py:
```diff
@@ -0,0 +1,218 @@
+        return errors.not_found("The item with the ID: {0} doesn't exist"
+                                .format(item_id))
+
+    if not request.json or (('name' and 'done') not in request.json):
```
why can't the user update only the `name` or `done`.. why do they have to send both..

BolajiOlajide replied:
I've resolved this.

njirap commented on app/main/routes.py:
```diff
@@ -0,0 +1,218 @@
+                                .format(item_id))
+
+    if not request.json or (('name' and 'done') not in request.json):
+        return errors.bad_request("No JSON file detected.")
```
is this the best error here?

BolajiOlajide replied:
I've resolved this issue.

njirap commented on app/models.py:
```diff
@@ -0,0 +1,286 @@
+The SQLAlchemy models for the database is defined here.
+"""
+
+import datetime
```
maybe `from datetime import datetime`

BolajiOlajide replied:
Awesome. I've corrected that also.

njirap commented on config.py:
```diff
@@ -0,0 +1,90 @@
+        This static method initializes the application using whatever
+        configuration the user has chosen.
+        """
+        pass
```
Why have a function whose purpose is to `pass`?

BolajiOlajide replied:
I have removed the irrelevant function.

njirap commented on shell.py:
```diff
@@ -0,0 +1,27 @@
+prompt.
+"""
+import os
+from os.path import join, dirname
```
maybe add `os.environ` here and remove L7

BolajiOlajide replied:
I have fixed that using the statement `from os import environ`

njirap commented on tests/test_app.py:
```diff
@@ -0,0 +1,59 @@
+
+Test the basic settings of the application to be certain it's functioning well.
+"""
+import unittest
```
suggestion.. maybe use `flask_testing`

BolajiOlajide replied:
I'll look into it.

njirap commented on tests/test_config.py:
```diff
@@ -0,0 +1,136 @@
+Test the configuration of the application to be certain it's functioning well.
+"""
+from os.path import join
+import os
```
merge this with L6 imports..

BolajiOlajide replied:
I've corrected it.

njirap commented on tests/test_config.py:
```diff
@@ -0,0 +1,136 @@
+        Test that DEBUG and SQLALCHEMY_TRACK_MODIFICATIONS are set to True and
+        the database used is the local one.
+        """
+        self.assertTrue(self.app.config['DEBUG'] is True)
```
`self.assertTrue(self.app.config['DEBUG'])`

BolajiOlajide replied:
Corrected.

njirap commented on tests/test_config.py:
```diff
@@ -0,0 +1,136 @@
+        the database used is the local one.
+        """
+        dbase = os.environ.get("TEST_DB")
+        self.assertFalse(self.app.config['USE_RATE_LIMITS'] is True)
```
`self.assertFalse(self.app.config['USE_RATE_LIMITS'])`

BolajiOlajide replied:
Corrected.

njirap commented on tox.ini:
```diff
@@ -0,0 +1,8 @@
+commands=python manage.py test
+
+[testenv:py27]
+basepython=python2.7
```
maybe add more tests like `pep8`

BolajiOlajide replied:
I'll do that. Thanks

mfalade commented on app/__init__.py:
```diff
@@ -0,0 +1,33 @@
+    """
+    app = Flask(__name__)
+    app.config.from_object(config[config_name])
+    config[config_name].init_app(app)
```
You don't need to do this. I'd suggest you remove the method definition in your config.py.

BolajiOlajide replied:
Hi, i've removed that line of code.

BolajiOlajide committed:
Implement Feedback from my Facilitator (#3)

* chore: delete migrations

* Ignore coverage and sqlite database

* chore: correct spelling error

* remove redundant function call

* rf: refactor code

* delete database

* add flake8 to requirements

* remove unused imports

* fix incorrect status code for items created

* configure tox for flake8 testing

* add migration files

* chore: exclude migration files for flake8 test

* chore: remove redundant test

* ft: implement mixin for models

* specify error messages for errors

BolajiOlajide committed:
configure cors
